src/remoted/secure.c

```diff
@@ -629,49 +625,54 @@ STATIC void HandleSecureMessage(const message_t *message, int *wdb_sock) {
     /* Check if it is a control message */
     if (IsValidHeader(tmp_msg)) {
 
-        /* We need to save the peerinfo if it is a control msg */
+        /* let through new and shutdown messages */
+        if (message->counter >  rem_getCounter(message->sock) || (strncmp(tmp_msg, HC_SHUTDOWN, strlen(HC_SHUTDOWN)) == 0)) {
+            /* We need to save the peerinfo if it is a control msg */
 
-        w_mutex_lock(&keys.keyentries[agentid]->mutex);
-        keys.keyentries[agentid]->net_protocol = protocol;
-        keys.keyentries[agentid]->rcvd = time(0);
-        memcpy(&keys.keyentries[agentid]->peer_info, &message->addr, logr.peer_size);
+            w_mutex_lock(&keys.keyentries[agentid]->mutex);
+            keys.keyentries[agentid]->net_protocol = protocol;
+            keys.keyentries[agentid]->rcvd = time(0);
+            memcpy(&keys.keyentries[agentid]->peer_info, &message->addr, logr.peer_size);
 
-        keyentry * key = OS_DupKeyEntry(keys.keyentries[agentid]);
+            keyentry * key = OS_DupKeyEntry(keys.keyentries[agentid]);
 
-        if (protocol == REMOTED_NET_PROTOCOL_TCP) {
-            if (message->counter > rem_getCounter(message->sock)) {
-                keys.keyentries[agentid]->sock = message->sock;
-            }
+            if (protocol == REMOTED_NET_PROTOCOL_TCP) {
+                if (message->counter > rem_getCounter(message->sock)) {
+                    keys.keyentries[agentid]->sock = message->sock;
+                }
 
-            w_mutex_unlock(&keys.keyentries[agentid]->mutex);
+                w_mutex_unlock(&keys.keyentries[agentid]->mutex);
 
-            r = OS_AddSocket(&keys, agentid, message->sock);
-
-            switch (r) {
-            case OS_ADDSOCKET_ERROR:
-                merror("Couldn't add TCP socket to keystore.");
-                break;
-            case OS_ADDSOCKET_KEY_UPDATED:
-                mdebug2("TCP socket %d already in keystore. Updating...", message->sock);
-                break;
-            case OS_ADDSOCKET_KEY_ADDED:
-                mdebug2("TCP socket %d added to keystore.", message->sock);
-                break;
-            default:
-                ;
+                r = OS_AddSocket(&keys, agentid, message->sock);
```

Comment:
I think this shouldn't be done when the message is a shutdown because the socket was probably removed from the `rb_tree` at this point.

Reply:
Not at all shutdown messages the socket will be closed. Most of the time it is still open.